2022/Q2/artifacts/pos-model/PoS-model.md

```diff
@@ -69,35 +78,21 @@ total_voting_power[] in Epoch to VotingPower //map from epoch to voting_power
 ```go
 tx_become_validator(validator_address, consensus_key, staking_reward_address)
 {
-  //VP:https://github.com/anoma/anoma/blob/master/proof_of_stake/src/validation.rs#L421
-  //https://github.com/anoma/anoma/blob/master/proof_of_stake/src/validation.rs#L459
-  //https://github.com/anoma/anoma/blob/master/proof_of_stake/src/validation.rs#L471
-  //COMMENT: In the code they check not only at cur_epoch but at all epochs between cur_epoch and the offset. Double-check
-  //COMMENT: I do not see in the code where they check that the consensus key is not an old one (this is in the docs).
-  //COMMENT: todo: check there are no consensus_key changes schedule between cur_epoch and the offset. If this necessary though? should
-  //not be enough with the state checks?
-  //https://github.com/anoma/anoma/blob/master/proof_of_stake/src/validation.rs#L515
-  var pre_state = read_epoched_field(validators[validator_address].state, cur_epoch)
-  if (pre_state == ⊥ && validator_address != staking_reward_address) then
-    //reward_address is not in the docs/spec validator struct
+  //check that become_validator has not been called before for validator_address
+  var cur_state = read_epoched_field(validators[validator_address].state, cur_epoch+pipeline_length)
+  if (cur_state == ⊥ && validator_address != staking_reward_address) then
     validators[validator_address].reward_address = staking_reward_address
-    //set status to pending inmediately
-    validators[validator_address].state[cur_epoch] = pending
     //set status to candidate and consensus key at n + pipeline_length
     validators[validator_address].consensus_key[cur_epoch+pipeline_length] = consensus_key
```

Comment:
So now we do not set the status to anything between `cur_epoch` and `cur_epoch+pipeline_length`?

Reply:
That's right. This is the conclusion we got at in the meeting with Tomas this week.
Rationale is: there's nothing we can do to change the _actual_ state of the current epoch, so changing it in the store introduces a mismatch (and confusion).
And (my own thought here): setting the current epoch to _pending_ does not add any information. If you are _inactive_ (or _bottom_) in the current epoch and you set `cur_epoch+pipeline_length` to _candidate_, and observer of the history will conclude you are efectively in "pending" state in the current epoch.

Side effect: conditions got _so much simpler_ in all three transaction types! (`become_validator`, `deactivate`, `reactivate`)